lib/spack/spack/spec.py

```diff
@@ -1812,10 +1812,12 @@ def concretize(self):
                     for patch in patch_list:
                         patches.append(patch.sha256)
             if patches:
-                # Special-case: keeps variant values unique but ordered.
-                s.variants['patches'] = MultiValuedVariant('patches', ())
-                mvar = s.variants['patches']
-                mvar._value = mvar._original_value = tuple(dedupe(patches))
+                mvar = s.variants.setdefault(
+                    'patches', MultiValuedVariant('patches', ())
+                )
+                mvar.value = patches
```

Comment:
Is there a reason this no longer needs `dedupe`?

Reply:
Yes. The statement:
```python
m.value = patches
```
activates [this setter](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/variant.py#L268) that internally does:
```python
m._value = tuple(sorted(set(patches)))
```
(note the underscore above) making `dedupe` unnecessary here. 

If I interpret correctly @tgamblin intents, setting the "private" value using `dedupe` was meant to preserve the order of patches, i.e. the correct application order. What we didn't take into account is that the ordering of patches right now is weak (enforced among common constraints) and `patches` is populated iterating over a dictionary (that doesn't preserve order).

The nastiness in this PR is in the last commit, where I set an attribute **on the instance** of  `MultiValuedVariant` used to store patches. That attribute stores the order in which good patches appeared. Anyhow, it's contained in 3 statements and a single module (`spec.py`)